src/libraries/System.Private.CoreLib/src/System/Globalization/InvariantModeCasing.cs

```diff
@@ -0,0 +1,369 @@
+                {
+                    char cl = source[i + 1];
+                    if (char.IsLowSurrogate(cl))
+                    {
+                        // well formed surrogates
+                        SurrogateCasing.ToUpper(c, cl, out char h, out char l);
+                        destination[i]   = h;
+                        destination[i+1] = l;
+                        i++; // skip the low surrogate
+                        continue;
+                    }
+                }
+
+                destination[i] = ToUpper(c);
+            }
+        }
+
+        internal static void ToLower(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(GlobalizationMode.Invariant);
+            Debug.Assert(source.Length <= destination.Length);
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                char c = source[i];
+                if (char.IsHighSurrogate(c) && i < source.Length - 1 )
+                {
+                    char cl = source[i + 1];
+                    if (char.IsLowSurrogate(cl))
+                    {
+                        // well formed surrogates
+                        SurrogateCasing.ToLower(c, cl, out char h, out char l);
+                        destination[i]   = h;
+                        destination[i+1] = l;
+                        i++; // skip the low surrogate
+                        continue;
+                    }
+                }
+
+                destination[i] = ToLower(c);
+            }
+        }
+
+        internal static int CompareStringIgnoreCase(ref char strA, int lengthA, ref char strB, int lengthB)
+        {
+            Debug.Assert(GlobalizationMode.Invariant);
+
+            int length = Math.Min(lengthA, lengthB);
+
+            ref char charA = ref strA;
+            ref char charB = ref strB;
+
+            while (length != 0)
+            {
+                if (!char.IsHighSurrogate(charA) || !char.IsHighSurrogate(charB))
+                {
+                    if (charA == charB)
+                    {
+                        length--;
+                        charA = ref Unsafe.Add(ref charA, 1);
+                        charB = ref Unsafe.Add(ref charB, 1);
+                        continue;
+                    }
+
+                    char aUpper = ToUpper(charA);
+                    char bUpper = ToUpper(charB);
+
+                    if (aUpper == bUpper)
+                    {
+                        length--;
+                        charA = ref Unsafe.Add(ref charA, 1);
+                        charB = ref Unsafe.Add(ref charB, 1);
+                        continue;
+                    }
+
+                    return aUpper - bUpper;
+                }
+
+                if (length == 1)
+                {
+                    return charA - charB;
+                }
+
+                // We come here only if we have valid high surrogates and length > 1
+
+                char a = charA;
```

Comment:
Time-of-check/time-of-use: _a_ and _b_ are not guaranteed to be high surrogate chars at this point, as the underlying contents may have changed between the original check on line 173 above. (The underlying contents could be an arbitrary `ROS<char>`, not a known-immutable `string`.)

Reply:
I already changed this code.